qutebrowser/misc/sql.py

```diff
@@ -358,10 +363,58 @@ def insert_batch(self, values, replace=False):
             values: A dict with a list of values to insert for each field name.
             replace: If true, overwrite rows with a primary key match.
         """
-        q = self._insert_query(values, replace)
+        q = Query(self._insert_query(values, replace))
         q.run_batch(values)
         self.changed.emit()
 
+    def update(self, update, where, escape=True):
+        """Execute update rows statement.
+
+        Args:
+            update: column:value dict with new values to set
+            where: column:value dict for filtering
+            escape: enable new values SQL-escaping
+        """
+        if escape:
+            u = ', '.join(('{} = :{}'.format(k, k) for k in update.keys()))
+        else:
+            u = ', '.join(('{} = {}'.format(k, v) for k, v in update.items()))
+
+        s = 'UPDATE {} SET {}'.format(self._name, u)
+        if where:
+            w = ' AND '.join(('{} = :w_{}'.format(f, f) for f in where.keys()))
+            s += ' WHERE {}'.format(w)
+
+        p = update.copy()
+        p.update({'w_' + k: v for k, v in where.items()})
+
+        Query(s).run(**p)
+        self.changed.emit()
+
+    def upsert(self, values, index, update, escape=True):
+        """
+        Insert or update row if it is alredy exists.
+
+        Args:
+            update: column:value dict for insert operation
+            index: column name used as a primary index
+            update: column:value dict for update operation
+            escape: enable SQL-escaping for update operation
+        """
+
+        # TODO: ON CONFLICT can be used after updating to SQLite v3.24.0.
+        #       Something like
+        #       q = self._insert_query(values, False)
+        #       q += ' ON CONFLICT ({}) DO UPDATE SET {}'.format(index, update)
+
+        q = Query(self._insert_query(values, ignore=True))
+        q.run(**values)
+
+        if not q.rows_affected() and update:
```

Comment:
Since we have index on `url` column this update is very fast. SQLite does this kind of update on my machine in less then millisecond on table with million records.

Speaking about hitting DB on page load. I noticed that we add more then one history record for the single page (which may distort visits/frecency number a little. Looks like not critical).
 I have not dug too much for this, my plan is to get back to it after the current tasks. If should prevent unnecessary inserts.

Reply:
> I noticed that we add more then one history record for the single page

This is a known issue: https://github.com/qutebrowser/qutebrowser/issues/4310